modules/ip6/datapath/gr_ip6_datapath.h

```diff
@@ -17,8 +17,63 @@
 
 #include <stdint.h>
 
-GR_MBUF_PRIV_DATA_TYPE(ip6_output_mbuf_data, { struct nexthop6 *nh; });
+GR_MBUF_PRIV_DATA_TYPE(ip6_output_mbuf_data, {
+	const struct iface *input_iface;
+	struct nexthop6 *nh;
+});
+
+GR_MBUF_PRIV_DATA_TYPE(ip6_local_mbuf_data, {
+	struct rte_ipv6_addr src;
+	struct rte_ipv6_addr dst;
+	uint16_t len;
+	uint8_t hop_limit;
+	uint8_t proto;
+	const struct iface *input_iface;
+});
+
+void ip6_input_local_add_proto(uint8_t proto, const char *next_node);
 
 #define IP6_VTC_FLOW UINT32_C(0x60000000)
+#define IP6_DEFAULT_HOP_LIMIT 255
+
+static inline void ip6_set_fields(
+	struct rte_ipv6_hdr *ip,
+	uint16_t len,
+	uint8_t proto,
+	struct rte_ipv6_addr *src,
+	struct rte_ipv6_addr *dst
+) {
+	ip->vtc_flow = RTE_BE32(IP6_VTC_FLOW);
+	ip->payload_len = rte_cpu_to_be_16(len);
+	ip->proto = proto;
+	ip->hop_limits = IP6_DEFAULT_HOP_LIMIT;
+	rte_ipv6_addr_cpy(&ip->src_addr, src);
+	rte_ipv6_addr_cpy(&ip->dst_addr, dst);
+}
+
+struct ip6_pseudo_header {
+	rte_be32_t payload_len;
+	uint16_t __zero1;
+	uint8_t __zero2;
+	uint8_t proto;
+} __rte_packed;
+
+static inline uint16_t ip6_cksum(
```

Comment:
Why not reuse rte_ipv6_phdr_cksum?

Reply:
Because it takes a ipv6_hdr_addr in parameter. Which I don't have yet when constructing icmp packets. Also, I think rte_ipv6_phdr_cksum only works on little endian machines. I'll check if I can do something about that.